File: mobile/lib/services/crypto/key_manager.dart
import 'dart:convert';

import 'package:cryptography/cryptography.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

/// Gestionnaire de la paire de clés Ed25519 du téléphone Linkup.
///
/// - Génère une paire au premier appel à `ensureKeyPair()`
/// - Stocke la **seed 32 octets** (base64) dans le secure storage Android
///   (Keystore matériel) sous deux clés : public + seed
/// - Re-dérive la `SimpleKeyPair` depuis la seed à chaque lecture
///
/// S2.J1 — base pour le handshake Noise IK (S2.J3) et la signature des
/// messages broadcast Reverb (S3).
class KeyManager {
  static const String _seedKey = 'linkup.ed25519.seed';
  static const String _publicKey = 'linkup.ed25519.public';

  final FlutterSecureStorage _storage;
  final Ed25519 _algo;

  KeyManager({
    FlutterSecureStorage? storage,
    Ed25519? algo,
  })  : _storage = storage ?? const FlutterSecureStorage(),
        _algo = algo ?? Ed25519();

  /// Renvoie la paire de clés, génère + stocke si absente.
  Future<SimpleKeyPair> ensureKeyPair() async {
    final existing = await _loadKeyPair();
    if (existing != null) return existing;
    return _generateAndStore();
  }

  /// Génère une nouvelle paire et l'écrase en stockage. Réservé aux tests
  /// ou à une rotation explicite — sinon préférer ensureKeyPair().
  Future<SimpleKeyPair> generate() async {
    return _generateAndStore();
  }

  /// Clé publique base64 — utilisée dans le QR pairing.
  Future<String> publicKeyBase64() async {
    final pub = await _readPublicKey();
    return base64.encode(pub.bytes);
  }

  /// Signe un message avec la clé privée locale.
  /// Retourne la signature détachée en base64.
  Future<String> sign(List<int> message) async {
    final kp = await ensureKeyPair();
    final signature = await _algo.sign(message, keyPair: kp);
    return base64.encode(signature.bytes);
  }

  /// Vérifie une signature contre une clé publique base64 (signature détachée).
  ///
  /// Réservé S3 : pas encore appelé dans `lib/` (le pairing valide l'identité
  /// du PC par égalité de clé publique, pas par signature). Servira à vérifier
  /// les messages broadcast Reverb signés par le PC. Couvert par les tests.
  Future<bool> verify({
    required List<int> message,
    required String signatureB64,
    required String publicKeyB64,
  }) async {
    try {
      final signature = Signature(
        base64.decode(signatureB64),
        publicKey: SimplePublicKey(
          base64.decode(publicKeyB64),
          type: KeyPairType.ed25519,
        ),
      );
      return _algo.verify(message, signature: signature);
    } on FormatException {
      return false;
    }
  }

  /// Empreinte SHA-256 courte (8 hex chars) de la clé publique du tél.
  ///
  /// Réservé : pas encore appelé dans `lib/` (l'empreinte du tél affichée vient
  /// de la réponse serveur `device_fingerprint`). Gardé pour un affichage local
  /// futur de sa propre empreinte. Couvert par les tests. Même algo que le PC
  /// (`KeyManager::fingerprint` côté Laravel) → valeurs comparables.
  Future<String> fingerprint() async {
    final pub = await _readPublicKey();
    final hash = await Sha256().hash(pub.bytes);
    return hash.bytes
        .sublist(0, 4)
        .map((b) => b.toRadixString(16).padLeft(2, '0'))
        .join();
  }

  /// Vrai si une paire est déjà stockée.
  Future<bool> exists() async {
    final seed = await _storage.read(key: _seedKey);
    final pub = await _storage.read(key: _publicKey);
    return seed != null && pub != null;
  }

  /// Efface la paire stockée. Utile pour les tests ou un reset complet.
  Future<void> clear() async {
    await _storage.delete(key: _seedKey);
    await _storage.delete(key: _publicKey);
  }

  Future<SimpleKeyPair> _generateAndStore() async {
    final kp = await _algo.newKeyPair();
    final seed = await kp.extractPrivateKeyBytes();
    final pub = await kp.extractPublicKey();
    await _storage.write(key: _seedKey, value: base64.encode(seed));
    await _storage.write(key: _publicKey, value: base64.encode(pub.bytes));
    return kp;
  }

  Future<SimpleKeyPair?> _loadKeyPair() async {
    final seedB64 = await _storage.read(key: _seedKey);
    if (seedB64 == null) return null;
    try {
      final seed = base64.decode(seedB64);
      return _algo.newKeyPairFromSeed(seed);
    } on FormatException {
      return null;
    }
  }

  Future<SimplePublicKey> _readPublicKey() async {
    final kp = await ensureKeyPair();
    return kp.extractPublicKey();
  }
}

File: agent/app/Services/Crypto/KeyManager.php
<?php

namespace App\Services\Crypto;

use RuntimeException;

/**
 * Gestionnaire de la paire de clés Ed25519 de l'agent Linkup (PC).
 *
 * - Génère une paire au premier appel à `ensureKeyPair()`
 * - Stocke base64 dans `~/.linkup/keys/agent_ed25519.{pub,sec}`
 * - Le `.sec` est chmod 600 (lisible seulement par le user owner)
 * - Le dossier `~/.linkup/keys/` est chmod 700
 *
 * S2.J1 — base pour le handshake Noise IK (S2.J3) et la signature des
 * messages broadcast Reverb (S3).
 */
class KeyManager
{
    private const KEY_DIR = '.linkup/keys';
    private const PUB_FILE = 'agent_ed25519.pub';
    private const SEC_FILE = 'agent_ed25519.sec';

    public function __construct(
        private readonly string $homeDir,
    ) {
    }

    /**
     * Garantit qu'une paire existe ; la génère sinon.
     * @return array{public: string, secret: string} (base64)
     */
    public function ensureKeyPair(): array
    {
        if ($this->exists()) {
            return $this->load();
        }
        return $this->generate();
    }

    /**
     * Génère une nouvelle paire Ed25519 et l'écrit sur disque.
     * Écrase toute paire existante (rarement souhaitable — utiliser ensureKeyPair).
     */
    public function generate(): array
    {
        $kp = sodium_crypto_sign_keypair();
        $public = base64_encode(sodium_crypto_sign_publickey($kp));
        $secret = base64_encode(sodium_crypto_sign_secretkey($kp));

        $this->ensureDir();
        file_put_contents($this->pubPath(), $public);
        file_put_contents($this->secPath(), $secret);
        chmod($this->secPath(), 0600);

        return ['public' => $public, 'secret' => $secret];
    }

    /**
     * Charge la paire existante. Throws si absente.
     */
    public function load(): array
    {
        if (!$this->exists()) {
            throw new RuntimeException(
                "Paire Ed25519 introuvable dans {$this->keyDirPath()}. "
                . 'Appeler ensureKeyPair() pour la générer.'
            );
        }
        return [
            'public' => trim((string) file_get_contents($this->pubPath())),
            'secret' => trim((string) file_get_contents($this->secPath())),
        ];
    }

    /** Renvoie juste la clé publique base64 — utile pour le QR pairing. */
    public function publicKey(): string
    {
        return $this->ensureKeyPair()['public'];
    }

    /**
     * Signe un message avec la clé privée locale.
     * Retourne la signature détachée en base64.
     */
    public function sign(string $message): string
    {
        $secret = base64_decode($this->ensureKeyPair()['secret'], true);
        if ($secret === false) {
            throw new RuntimeException('Clé secrète corrompue (base64 invalide).');
        }
        return base64_encode(sodium_crypto_sign_detached($message, $secret));
    }

    /**
     * Vérifie une signature contre une clé publique (base64).
     * Utile pour valider des messages venant d'un device pairé.
     */
    public function verify(string $message, string $signatureB64, string $publicKeyB64): bool
    {
        $signature = base64_decode($signatureB64, true);
        $publicKey = base64_decode($publicKeyB64, true);
        if ($signature === false || $publicKey === false) {
            return false;
        }
        try {
            return sodium_crypto_sign_verify_detached($signature, $message, $publicKey);
        } catch (\SodiumException) {
            return false;
        }
    }

    /** Empreinte SHA-256 courte (8 hex chars) de la clé publique locale — affichage QR popup. */
    public function fingerprint(): string
    {
        return $this->fingerprintOf($this->publicKey());
    }

    /**
     * Empreinte SHA-256 courte (8 hex chars) d'une clé publique Ed25519 base64.
     *
     * Calcul centralisé du fingerprint (évite la duplication avec
     * PairingHandshakeService, qui délègue ici). Throws si base64 invalide.
     */
    public function fingerprintOf(string $publicKeyBase64): string
    {
        $raw = base64_decode($publicKeyBase64, true);
        if ($raw === false) {
            throw new RuntimeException('Clé publique non décodable (base64 invalide).');
        }
        return substr(hash('sha256', $raw), 0, 8);
    }

    public function exists(): bool
    {
        return is_file($this->pubPath()) && is_file($this->secPath());
    }

    private function ensureDir(): void
    {
        $dir = $this->keyDirPath();
        if (!is_dir($dir)) {
            if (!mkdir($dir, 0700, true) && !is_dir($dir)) {
                throw new RuntimeException("Impossible de créer $dir");
            }
        }
        chmod($dir, 0700);
    }

    private function keyDirPath(): string
    {
        return rtrim($this->homeDir, '/') . '/' . self::KEY_DIR;
    }

    private function pubPath(): string
    {
        return $this->keyDirPath() . '/' . self::PUB_FILE;
    }

    private function secPath(): string
    {
        return $this->keyDirPath() . '/' . self::SEC_FILE;
    }
}
